vsp.go

```diff
@@ -57,65 +61,316 @@ func (mw *MultiWallet) NewVSPClient(vspHost string, walletID int, purchaseAccoun
+				cancels = cancels[:0]
+
+				intervalSize := int32(w.ChainParams().StakeDiffWindowSize)
+				currentInterval := height / intervalSize
+				nextIntervalStart = (currentInterval + 1) * intervalSize
+
+				// Skip this purchase when no more tickets may be purchased in the interval and
+				// the next sdiff is unknown.  The earliest any ticket may be mined is two
+				// blocks from now, with the next block containing the split transaction
+				// that the ticket purchase spends.
+				if height+2 == nextIntervalStart {
+					log.Debugf("Skipping purchase: next sdiff interval starts soon")
+					continue
+				}
+				// Set expiry to prevent tickets from being mined in the next
+				// sdiff interval.  When the next block begins the new interval,
+				// the ticket is being purchased for the next interval; therefore
+				// increment expiry by a full sdiff window size to prevent it
+				// being mined in the interval after the next.
+				expiry = nextIntervalStart
+				if height+1 == nextIntervalStart {
+					expiry += intervalSize
+				}
+			}
+
+			wal := vsp.w
+			// Get the account balance to determine how many tickets to buy
+			bal, err := wal.GetAccountBalance(int32(vsp.purchaseAccount))
+			if err != nil {
+				return err
+			}
+
+			spendable := bal.Spendable
+			if spendable < balanceToMaintain {
+				log.Debugf("Skipping purchase: low available balance on Wallet: %s", wal.Name)
+				return nil
+			}
+
+			spendable -= balanceToMaintain
+			sdiff, err := wal.Internal().NextStakeDifficultyAfterHeader(ctx, tipHeader)
+			if err != nil {
+				return err
+			}
+
+			buy := int(dcrutil.Amount(spendable) / sdiff)
+			if buy == 0 {
+				log.Debugf("Skipping purchase: low available balance on Wallet: %s", wal.Name)
+				return nil
+			}
+
+			max := int(wal.Internal().ChainParams().MaxFreshStakePerBlock)
+			if buy > max {
+				log.Infof("Ticket buyer is about to purchase %v tickets which is more the max ticket of %v per block.", buy, max)
+			}
+
+			cancelCtx, cancel := context.WithCancel(ctx)
+			cancels = append(cancels, cancel)
+			buyTickets := func() {
+				err := vsp.buyTickets(cancelCtx, passphrase, sdiff, expiry)
+				if err != nil {
+					switch {
+					// silence these errors
+					case errors.Is(err, errors.InsufficientBalance):
+					case errors.Is(err, context.Canceled):
+					case errors.Is(err, context.DeadlineExceeded):
+					default:
+						log.Errorf("[%d] Ticket purchasing failed: %v", wal.ID, err)
+					}
+					if errors.Is(err, errors.Passphrase) {
+						fatalMu.Lock()
+						fatal = err
+						fatalMu.Unlock()
+						outerCancel()
+					}
+				}
+			}
+
+			// start separate ticket purchase for as many tickets that can be purchased
+			// each purchase only buy 1 ticket.
+			for i := 0; i < buy; i++ {
+				go buyTickets()
+			}
+		}
+	}
+}
+
+// buyTickets purchases one or more tickets depending on the spendable balance of
+// the selected account, the specified minimum balance to maintain and the limit
+// per block.
```

Comment:
There's no more limit per block.
```suggestion
// the selected account and the specified minimum balance to maintain.
```